### ISSUE TITLE
test/fixedbugs/issue14646.go fails for sparc64 SSA backend

### ISSUE THREAD

**Issue by binarycrusader** [4ad/go#34]
```
# go run run.go -- fixedbugs/issue14646.go
incorrect output
Expected line = 18 but got line = 17 and file = /builds/srwalker/golang/go-sparc/test/fixedbugs/issue14646.go

FAIL    fixedbugs/issue14646.go 0.388s
```

**Comment by binarycrusader** [comment]
The root cause of this issue is the same as #29.

**Comment by binarycrusader** [comment]
This isn't the right fix, but the pc calculation is off for some reason for Caller/getcallerpc on sparc64, and this makes the test pass:

```patch
diff --git a/src/runtime/extern.go b/src/runtime/extern.go
index e09c3d0fd0..8b673ba4cb 100644
--- a/src/runtime/extern.go
+++ b/src/runtime/extern.go
@@ -190,7 +190,9 @@ func Caller(skip int) (pc uintptr, file string, line int, ok bool) {
        // If we see a call to sigpanic, we do not back up the PC to find
        // the line number of the call instruction, because there is no call.
        // SPARC64's PC holds the address of the *current* instruction.
-       if xpc > f.entry && (g == nil || g.entry != funcPC(sigpanic)) && sys.GoarchSparc64 == 0 {
+       if sys.GoarchSparc64 != 0 {
+               xpc += 4
+       } else if xpc > f.entry && (g == nil || g.entry != funcPC(sigpanic)) {
                xpc--
        }
        file, line32 := funcline(f, xpc)
```

The good thing at least is that this is also broken for the non-SSA backend, so this is something that has always needed fixing.

**Comment by 4ad** [comment]
Hmm, probably NPC vs PC confusion.

**Comment by binarycrusader** [comment]
The PC calculated here (before my +=4 hack) is the nop of the return from the deferred function (main.main.func1+0xa8):

```asm
main.main.func1+0xa4:                 retl
main.main.func1+0xa8:                 nop
main.main.func1+0xac:                 nop
main.main.func1+0xb0:                 call      +0x4d740      <runtime.deferreturn>
```

...so line 17 is an accurate answer.

In the SSA-backend case, my wild guess initially was that  this is related to the nop that's inserted by the SSA backend for OpSPARC64CALLstatic:

```go
case ssa.OpSPARC64CALLstatic:
        if v.Aux.(*gc.Sym) == gc.Deferreturn.Sym {
                // TODO(shawn): is this true on sparc due to pc/npc difference?
                // Deferred calls will appear to be returning to
                // the CALL deferreturn(SB) that we are about to emit.
                // However, the stack trace code will show the line
                // of the instruction byte before the return PC.
                // To avoid that being an unrelated instruction,
                // insert an actual hardware NOP that will have the right line number.
                // This is different from obj.ANOP, which is a virtual no-op
                // that doesn't make it into the instruction stream.
                ginsnop()
        }
```

However, it doesn't appear to be that simple.

**Comment by binarycrusader** [comment]
For a long time, we've attempted to "fix" gentraceback() and related trace functions by having them account for the fact that pc is the last executed instruction on sparc64 instead of the next instruction to execute.  Instead of doing that, we now advance pc by sys.PCQuantum when doing an initial trace to make it match other platforms, and we now correctly account for the fact that LR on sparc is not the actual return address -- 8(OLR) is, so we add sys.PCQuantum*2 whenever pc is derived from LR and we're not panicing.  In the case that we are panicing, the last instruction executed is the correct return address, at least for tracing purposes, such as determining lineno.

In addition, Go was wrongly doing "pc--" in several places when attempting to go the previous instruction, instead of subtracting sys.PCQuantum.  This worked on x86, etc. or any architecture where PCQuantum == 1, but it's unclear to me how it would ever work on any other arch, and it definitely doesn't work correctly on sparc64.

Fixed.
